paderwasn/rir_simulation/rir_simulation.py

```diff
@@ -0,0 +1,525 @@
+            # early reflections, so re-draw alpha, that then creates a lower,
+            # but achievable target DRR and therefore a solution during
+            # later scaling can be found
+            if max_factor < lower_limit_alpha:
+                # use omnidirectional directivity factor (1) as lower limit
+                # because otherwise no solution can be found
+                alpha = np.random.uniform(1, max_factor)
+            else:
+                # sample alpha until valid solution found, that still lays
+                # within the original range, but preserves convergence
+                alpha = np.random.uniform(lower_limit_alpha, max_factor)
+                while alpha > max_factor:
+                    alpha = np.random.uniform(lower_limit_alpha, max_factor)
+
+    target_drr = calc_target_drr(room_dimensions, alpha, beta, t60,
+                                 directivity_attenuation, distance)
+    weighting_function =\
+        calc_window(delta, kappa, toa_direct, rir_len, sample_rate)
+
+    scaling = calc_scaling(rir_ism, raw_stochastic, weighting_function,
+                           target_drr, upper_lim, lower_lim)
+    if scaling is None:
+        return None, alpha
+    else:
+        rir = rir_ism + raw_stochastic * weighting_function * scaling
+        return rir, alpha
+
+
+def calc_rirs(
+        src_pos, sensor_pos, room_dimensions, t60, azimuth_directivity,
+        colatitude_directivity, src_directivity,
+        directivity_factor_range=(2.5, 5.5), kappa=1, beta=1., rir_len=16384,
+        max_order_ism=3, win_len_direct=40, sound_speed=343, sample_rate=16000,
+        alpha_diff=0.1, max_retries=10
+):
+    """
+    Room impulse response (RIR) simulation for microphone pairs as described in
+    'DIMINISHING DOMAIN MISMATCH FOR DNN-BASED ACOUSTIC DISTANCE ESTIMATION
+    VIA STOCHASTIC ROOM REVERBERATION MODELS'
+
+    Args:
+        src_pos (array-like):
+            Coordinates of acoustic source with shape (dimension,)
+
+        sensor_pos (array-like):
+            Coordinates of the two microphones of the microphone pair with
+            shape (dimension, 2). It is assumed that the microphones have a
+            reasonably spacing for the assumnptions to hold.
+
+        room_dimensions (array-like):
+            Size of the room with shape (dimension)
+
+        t60 (float):
+            T60 time of the room in seconds
+
+        azimuth_directivity (float):
+            Azimuth of source in degrees
+
+        colatitude_directivity (float):
+            Colatitude of source in degrees
+
+        src_directivity (str):
+            Directivity pattern of the source; possible choices:
+            'cardioid', 'hypercardioid', 'supercardioid' 'subcardioid' and
+            'omnidirectional'
+
+        directivity_factor_range (array-like):
+            Range from which the directional factor of the sound source (alpha)
```

Comment:
Limits of the uniform distribution form which the directivity factor of the source (alpha) is drawn